### ISSUE TITLE
Memory leak when using Rpc server

### ISSUE THREAD

**Issue by pgScorpio** [jamulussoftware/jamulus#2614]
**Describe the bug**
CServerRpc and CClientRpc/CServerRpc  are created with "new" in main.cpp, but are never deleted.

**To Reproduce**
Run Jamulus with and without valid --jsonrpcport and --jsonrpcsecretfile parameters.
Use Visual studio's memory usage diagnostic tool to check the difference in memory usage at exit.

**Expected behavior**
Cleanup on exit.

**Screenshots**

**Operating system**
n.a.

**Version of Jamulus**
master @ 953f2b8a1ccab70d715a54cc730e05059ae140d6

**Additional context**

**Comment by hoffie** [comment]
Thanks for the report.
Without having looked at the code, these seem to be the handler instances, so they are only instantiated once on startup, right?

I vaguely remember that other init-once stuff was not properly cleaned up on exit in the past, but I'm not totally sure and don't remember if it might have been fixed meanwhile.

I think it makes sense to fix it in order to have a clean output when chasing more relevant leaks.

Are you planning to submit a PR for this issue? If so, that would be great and I'll assign it to you. It will hopefully be a non-controversial PR which can be merged quickly.

(I've edited the description to point to a master commit as 3.8.2 (final) can't have been affected as it didn't have the RPC code yet)

**Comment by dtinth** [comment]
@pgScorpio Hello, thanks for the report on memory leaking issue. I'm the original author of the JSON-RPC code.

IMO this issue may not be a high impact one because the RPC server is initialized once and used throughout the lifetime of the process. Still I agree that it is a memory leak nonetheless and it is better to have a cleaner output from memory profiler. Most of my programming background is in a garbage-collected language and so I forgot how to properly manage memory in C++ already 😂 

On my first implementation, I allocated an RPC server directly on the stack, but after code reviews, it has been changed to use C++ smart pointers, and subsequently to use just the `new` keyword. I'm not sure how to properly deallocate objects when it went out of scope without using smart pointers, so if you can contribute a proper fix, I would appreciate it. 😄

**Comment by pgScorpio** [comment]
@hoffie 

> I vaguely remember that other init-once stuff was not properly cleaned up on exit in the past, but I'm not totally sure and don't remember if it might have been fixed meanwhile.

Yes, there is still a lot more. i.e. ASIOSDK also has a big memory leak, and we can't change that, but that should be solved with the new Sound implementation (not using those ASIOSDK classes/functions.)
 
> I think it makes sense to fix it in order to have a clean output when chasing more relevant leaks.
> 
> Are you planning to submit a PR for this issue? If so, that would be great and I'll assign it to you. It will hopefully be a non-controversial PR which can be merged quickly.

Yes, I just did a pull request, See #2618

**Comment by pljones** [comment]
Moved to 3.9.1

**Comment by pljones** [comment]
Moving to 3.10.0.

**Comment by pljones** [comment]
Moving to Triage and removing milestone.

**Comment by pljones** [comment]
Closing as no activity in two years.

**Comment by mcfnord** [comment]
MY LLM WROTE:

**All three objects named here are already parented** — `CClientRpc`/`CServerRpc` to `pRpcServer` (main.cpp:993, :1061), `QTcpServer` to `this` (rpcserver.cpp:56), `pRpcServer` to `pApp` (main.cpp:969) — since @dtinth's original `d7047171`. The only unowned allocation is `pApp` itself (main.cpp:877/:884/:886), never deleted before `return 0` at :1127.

Headless server-only, `-g -O0`, valgrind memcheck, aarch64, at least 3 runs per point:

| | definitely lost | still reachable |
|---|---|---|
| RPC off | 0 | 19,760 B / 55 blocks |
| RPC on | 0 | 23,744 B / 116 blocks |

The RPC subsystem costs 3,984 B in 61 blocks, all reachable, none lost.

**It does not grow with use.** Through 500 connections closed three ways — normally, by the server via the batch-error path (rpcserver.cpp:156-162), and by RST abort — still reachable stays 23,776–23,848 B and the block count is **116 at every point**. The `disconnected` cleanup at rpcserver.cpp:125-130 fires on all three paths. One-shot at exit, which answers @hoffie's question and matches @dtinth's reasoning.

**`delete pApp;` before `return 0;`**, against #2618's +86/−16: exits 0 in 6/6 runs, no invalid read or write, no double free; still reachable drops to 18,024–18,096 B / 29 blocks and the RPC frames disappear. But definitely lost goes 0 → 16 B direct + 320 B indirect — the `CSignalHandler` `QGlobalStatic` singleton (signalhandler.cpp:157). You both wanted cleaner profiler output, and the one-liner moves 336 B *into* the category profilers flag, so the singleton needs handling too.

(Both builds also report `possibly lost: 336 B`. That is glibc pthread TLS, unrelated to Jamulus.)

Not tested: GUI builds, non-Linux.

**Comment by dtinth** [comment]
Last week, my agent also looked at the JSON-RPC implementation, and came to the same conclusion that the RPC server is properly parented so it should cause a memleak.

Per the OP's repro steps:

> "Use Visual studio's memory usage diagnostic tool to check the difference in memory usage at exit."

I think the way to check if this issue is still relevant is to run Jamulus _with_ and _without_ JSON-RPC. If the "the difference in memory usage at exit" are the same in both cases, then I think this issue can be closed as unrelated to JSON-RPC.

**Comment by mcfnord** [comment]
MY LLM WROTE:

@dtinth — ran your with/without-RPC test. Headless server, N=0 (enable RPC, start, clean exit), valgrind memcheck, aarch64, 3 reps each arm, bit-identical every rep:

| at exit | definitely lost | still reachable |
|---|---|---|
| RPC off | 0 | 19,760 B / 55 blk |
| RPC on | 0 | 23,744 B / 116 blk |

The difference isn't zero — enabling RPC adds 3,984 B / 61 blocks — but it's all **still-reachable** (the live `CServerRpc`/`QTcpServer`/`pRpcServer` tree, released at exit), and definitely-lost is 0 both ways. A "memory at exit" tool shows that delta, which is why it reads as a leak; it's live memory, not lost. So: close as not-a-leak.
